lib/git_worktree.rb

```diff
@@ -0,0 +1,327 @@
+    entry && entry[:type] == :tree
+  end
+
+  def read_file(path, commit_sha = nil)
+    read_entry(fetch_entry(path, commit_sha))
+  end
+
+  def read_entry(entry)
+    @repo.lookup(entry[:oid]).content
+  end
+
+  def directory?(path, commit_sha = nil)
+    entry = fetch_entry(path, commit_sha)
+    entry[:type] == :tree
+  end
+
+  def entries(path, commit_sha = nil)
+    tree = path.empty? ? lookup_commit_tree(commit_sha || @commit_sha) : get_tree(path, commit_sha)
+    tree.find_all.collect { |e| e[:name] }
+  end
+
+  def nodes(path, commit_sha = nil)
+    tree = path.empty? ? lookup_commit_tree(commit_sha || @commit_sha) : get_tree(path, commit_sha)
+    entries = tree.find_all
+    entries.each do |entry|
+      entry[:full_name] = File.join(@base_name, path, entry[:name])
+      entry[:rel_path] = File.join(path, entry[:name])
+    end
+  end
+
+  def save_changes(message, owner = :local)
+    cid = commit(message)
+    owner == :local ? lock { merge(cid) } : merge_and_push(cid)
+    true
+  end
+
+  def commit(message)
+    tree = @current_index.write_tree(@repo)
+    parents = @repo.empty? ? [] : [@repo.ref('refs/heads/master').target].compact
+    create_commit(message, tree, parents)
+  end
+
+  def file_attributes(fname)
+    walker = Rugged::Walker.new(@repo)
+    walker.sorting(Rugged::SORT_DATE)
+    walker.push(@repo.ref('refs/heads/master').target)
+    commit = walker.find { |c| c.diff(:paths => [fname]).size > 0 }
+    return {} unless commit
+    {:updated_on => commit.time.gmtime, :updated_by => commit.author[:name]}
+  end
+
+  def merge(commit, rebase = false)
+    mb = @repo.ref('refs/heads/master')
+    merge_index = mb ? @repo.merge_commits(mb.target, commit) : nil
+    if merge_index && merge_index.conflicts?
+      result = differences_with_master(commit)
+      raise MiqException::MiqGitConflicts, result
+    end
+    commit = rebase(commit, merge_index, mb ? mb.target : nil) if rebase
+    @repo.reset(commit, :soft)
+  end
+
+  def rebase(commit, merge_index, parent)
+    commit_obj = commit if commit.class == Rugged::Commit
+    commit_obj ||= @repo.lookup(commit)
+    Rugged::Commit.create(@repo, :author    => commit_obj.author,
+                                 :committer => commit_obj.author,
+                                 :message   => commit_obj.message,
+                                 :parents   => parent ? [parent] : [],
+                                 :tree      => merge_index.write_tree(@repo))
+  end
+
+  def merge_and_push(commit)
+    rebase = false
+    push_lock do
+      @saved_cid = @repo.ref('refs/heads/master').target.oid
+      merge(commit, rebase)
+      rebase = true
+      @repo.push(@remote_name, ['refs/heads/master'], :credentials => @cred)
+    end
+  end
+
+  def fetch
+    options = {:credentials => @cred}
+    ssl_no_verify_options(options)
+    @repo.fetch(@remote_name, options).tap do |_|
+      ENV.delete('GIT_SSL_NO_VERIFY') if ENV.key?('GIT_SSL_NO_VERIFY')
```

Comment:
@matthewd 
We might be getting repos from our own servers or from external servers, our servers might be using self signed certs which we don't want to verify.  We are trying to accommodate both of the scenarios below

(1) GIT_SSL_NO_VERIFY=true git clone https://internal_server
(2) git clone https://external_server